core/Pos/Core/Update/Util.hs

```diff
@@ -27,45 +33,78 @@ import           Instances.TH.Lift ()
+
+checkUpdatePayload
+    :: (HasConfiguration, MonadError Text m, Bi UpdateProposalToSign)
+    => UpdatePayload
+    -> m ()
+checkUpdatePayload it = do
+    -- Linter denies using foldables on Maybe.
+    -- Suggests whenJust rather than forM_.
+    --
+    --   ¯\_(ツ)_/¯
+    --
+    whenJust (upProposal it) checkUpdateProposal
+    forM_ (upVotes it) checkUpdateVote
+
+checkBlockVersionModifier
+    :: (MonadError Text m)
+    => BlockVersionModifier
+    -> m ()
+checkBlockVersionModifier BlockVersionModifier {..} = do
+    whenJust bvmMpcThd checkCoinPortion
+    whenJust bvmHeavyDelThd checkCoinPortion
+    whenJust bvmUpdateVoteThd checkCoinPortion
+    whenJust bvmUpdateProposalThd checkCoinPortion
+    whenJust bvmSoftforkRule checkSoftforkRule
+
+checkSoftforkRule
+    :: (MonadError Text m)
+    => SoftforkRule
+    -> m ()
+checkSoftforkRule SoftforkRule {..} = do
+    checkCoinPortion srInitThd
+    checkCoinPortion srMinThd
+    checkCoinPortion srThdDecrement
 
 -- | 'SoftforkRule' formatter which restricts type.
 softforkRuleF :: Format r (SoftforkRule -> r)
 softforkRuleF = build
 
-mkUpdateProposal
-    :: (HasConfiguration, Bi UpdateProposalToSign)
-    => BlockVersion
-    -> BlockVersionModifier
-    -> SoftwareVersion
-    -> HM.HashMap SystemTag UpdateData
-    -> UpAttributes
-    -> PublicKey
-    -> Signature UpdateProposalToSign
-    -> Either Text UpdateProposal
-mkUpdateProposal
-    upBlockVersion
-    upBlockVersionMod
-    upSoftwareVersion
-    upData
-    upAttributes
-    upFrom
-    upSignature = do
-        let toSign =
-                UpdateProposalToSign
-                    upBlockVersion
-                    upBlockVersionMod
-                    upSoftwareVersion
-                    upData
-                    upAttributes
-        unless (checkSig SignUSProposal upFrom toSign upSignature) $
-            Left "UpdateProposal: signature is invalid"
-        pure UnsafeUpdateProposal{..}
+checkUpdateVote
+    :: (HasConfiguration, MonadError Text m)
+    => UpdateVote
+    -> m ()
+checkUpdateVote it =
+    unless sigValid (throwError "UpdateVote: invalid signature")
+  where
+    sigValid = checkSig SignUSVote (uvKey it) (uvProposalId it, uvDecision it) (uvSignature it)
+
+checkUpdateProposal
+    :: (HasConfiguration, MonadError Text m, Bi UpdateProposalToSign)
+    => UpdateProposal
+    -> m ()
+checkUpdateProposal it = do
+    checkBlockVersionModifier (upBlockVersionMod it)
+    checkSoftwareVersion (upSoftwareVersion it)
+    forM_ (HM.keys (upData it)) checkSystemTag
+    let toSign = UpdateProposalToSign
+            (upBlockVersion it)
+            (upBlockVersionMod it)
+            (upSoftwareVersion it)
+            (upData it)
+            (upAttributes it)
+    unless (checkSig SignUSProposal (upFrom it) toSign (upSignature it))
+        (throwError "UpdateProposal: invalid signature")
+    checkSoftwareVersion (upSoftwareVersion it)
```

Comment:
You also `checkSoftwareVersion` twice.

Reply:
Ah, must have been a merge solution problem.